ARC-54 control analysis - PT1: run_all_scripts, run_all_control_scripts, run_crosstrack_analysis

MishkaMN opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description
This is Part 1 of control script analysis creation. This PR:
- added general `run_all_analysis` script that people can reuse by giving their custom analysis function (that has its own many performance metric analysis functions). Very common metric evaluation that we do in STOL is to do pass/no pass metrics based on the thresholds. Therefore, this function expects P/NP from each metric and can return summary for all overall files.
- Using that `run_all_analysis` script to add `run_all_control_analysis` script that houses all the different metric analysis such as cross_track error or spline fitting error etc. This is where the individual thresholds are defined and can be modifed to return P/NP
- Added cross_track analysis function to demonstrate how it works.
- Added unit test coverage for only `carma-platform` folder. This is so that the scripts are maintained and can be reused. Otherwise, we are creating too many new scripts and duplicating code from previous ROS1 approaches. If other projects don't desire to extend it, files can be excluded from it by adding them to `.sonarqube/sonar-project.properties`


Please see below pics and results to understand how the results might look like.
This PR is the first of many to make the PR digestible
I have more functions already implemented, which is partially in this old PR:  https://github.com/usdot-fhwa-stol/carma-analytics-fotda/pull/90

Design document for this task is here: https://leidoscorpus.sharepoint.us/:w:/r/sites/STR/TODevelopmentLibrary/22-256%20CDA%20Design%20%26%20Architecture/Work-In-Progress/Task%207%20Architecture%20Improvement%20Implementation/CARMA%20Platform%20Controls%20Improvements/Identified%20Core%20Performance%20Metrics%20And%20Analysis%20Design%20for%20CARMA%20Platform%20Control.docx?d=w40615c2fbfb74d439215cf62c5e5c4ca&csf=1&web=1&e=NXYkMT 

NOTE: very small amount of code is duplicate from carma1tenth. I am hoping once everything is done, that I can go back and reference this function in those scripts. I am hesitant to do it now as I might break those scripts, but I do believe this is the right place to have it in carma-platform as it is the "flagship" product. Please let me know if there is a better folder structure.

<!--- Describe your changes in detail -->

## Related GitHub Issue
NA
<!--- This project only accepts pull requests related to open GitHub issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key
https://usdot-carma.atlassian.net/browse/ARC-54
<!-- e.g. CAR-123 -->

## Motivation and Context
Control improvements under CDA DnA needed control performance analysis scripts.
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
Run simply by:
`python3 run_all_control_analysis.py --input-dir /workspaces/carma/carma_ws_ros2/src/analysis-data`
where `analysis-data` folder has all mcap files in different folders (rosbag2 folders):
![{026EB41C-15D0-494F-BF4A-03C6FE228D2F}](https://github.com/user-attachments/assets/e0d3dc30-175d-4288-9700-d068db4e6c04)

This will create an analysis folder that has overall results, numpy data, and plots for each mcap files:
![image](https://github.com/user-attachments/assets/437f181e-f970-4e69-b2fc-0a84181aa27a)


Resulting analysis_summary.json looks like this for example. NOTE: Some functions to generate such data are not in this PR:
```
{
  "analysis_time": "2024-10-25T06:52:52.306773",
  "analysis_type": "analysis",
  "total_files_analyzed": 2,
  "metrics_summary": {
    "run_crosstrack_analysis": {
      "total_files": 2,
      "passed": 1,
      "failed": 0,
      "errors": 1,
      "pass_rate": "50.00%",
      "error_rate": "50.00%"
    }
  },
  "analyzed_files": {
    "rosbag2_2024-10-22_213643_0.mcap": {
      "output_dir": "/workspaces/carma/src/analysis-data/analysis_20241025_065250/rosbag2_2024-10-22_213643_0",
      "metrics_results": {
        "run_crosstrack_analysis": true
      }
    },
    "rosbag2_2024_10_08-20_00_45_0.mcap": {
      "output_dir": "/workspaces/carma/src/analysis-data/analysis_20241025_065250/rosbag2_2024_10_08-20_00_45_0",
      "metrics_results": {
        "run_crosstrack_analysis": null
      }
    }
  }
}
```

As you can see `"rosbag2_2024_10_08-20_00_45_0.mcap` file is failing in all metrics because it doesn't have any meaningful topic data recorded.

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] I have added any new packages to the sonar-scanner.properties file
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

MishkaMN committed:
from personal PC

MishkaMN committed:
next

MishkaMN committed:
add run_all_script capability

MishkaMN committed:
Add spline based traj accuracy

MishkaMN committed:
More meaningful return files

MishkaMN committed:
Revert some changes to be digestable

MishkaMN committed:
add comment

MishkaMN committed:
remove extra results

MishkaMN commented on src/carma-platform/run_all_control_analysis.py:
```diff
@@ -0,0 +1,82 @@
+        except:
+            analysis_stats["run_crosstrack_analysis"] = None
+
+        # 2. Turn accuracy analysis by spline fitting
+        try:
+            stats, _, _, _ = run_turn_accuracy_analysis(mcap_path, data_dir, plots_dir)
+            analysis_stats["run_turn_accuracy_analysis"] = stats
+        except:
+            analysis_stats["run_turn_accuracy_analysis"] = None
+
+        # 3. Turn accuracy analysis by spline fitting
+        try:
+            stats, _, _, _ = run_acceleration_comfort_analysis(
+                mcap_path, data_dir, plots_dir
+            )
+            analysis_stats["run_acceleration_comfort_analysis"] = stats
+        except:
+            analysis_stats["run_acceleration_comfort_analysis"] = None
+
+        # 4. Turn accuracy analysis by spline fitting
+        try:
+            stats, _, _, _, _, _, _ = run_acceleration_comfort_analysis(
+                mcap_path, data_dir, plots_dir
+            )
+            analysis_stats["run_acceleration_comfort_analysis"] = stats
```
These don't exist yet, but just showing what it might look like. They will added to next PRs.

MishkaMN committed:
add pass or no pass functionality

MishkaMN committed:
reduce code by refactoring

MishkaMN committed:
more  documentatation

MishkaMN committed:
add readme

MishkaMN committed:
edit readme

MishkaMN committed:
add more readme

john-chrosniak commented on src/carma-platform/docs/cross_track_error_over_time.png:
It looks like the cross-track error analysis starts before the vehicle has started moving, which results in the median being the initial offset. Could we trim this based on when the vehicle starts/stops moving.

MishkaMN replied:
Great catch, I had totally forgotten. I added new function to get engage/disengage times and it looks better now.

john-chrosniak commented on src/carma-platform/parse_ros2_bags.py:
```diff
@@ -0,0 +1,121 @@
+from rclpy.serialization import deserialize_message
+
+
+def get_rosbag_options(path, serialization_format="cdr", storage_id="sqlite3"):
```
It looks like this logic is repeated in `open_bagfile`, so it should be fine to remove this

MishkaMN replied:
I just used the function instead of deleting in the `open_bagfile` since we factored it out anyways.

john-chrosniak commented on src/carma-platform/parse_ros2_bags.py:
```diff
@@ -0,0 +1,121 @@
+        raise ValueError(f"Missing field extractors for topics: {missing_extractors}")
+
+    # Open bag
+    reader, type_map = open_bagfile(str(mcap_path), topics=topics)
```
I'd also add a filter on the reader so we only loop through messages on the specified topics
Example: https://github.com/usdot-fhwa-stol/carma-analytics-fotda/blob/develop/src/carma-1-tenth/rosbag_utils.py#L25-L26

MishkaMN replied:
I thought the `open_bagfile` function already applied the filter. Is that correct?

john-chrosniak replied:
Ah yeah I missed that, thanks for pointing that out!

MishkaMN committed:
Add engage time

MishkaMN committed:
Arc 54 with unit tests (#92)

* add unit test

* fix typo

* add test_run_all_analysis

* Add sonar

* fix package

* add ros2-humble-key

* fix typo

* fix path

* fix path

* add humble source

* fix install

* update python

* add missing package yal

* add

* add pyYAML

* remove json

* add scipy

* fix location

* secret

* fix paths

* fix paths

* debug

* path

* Revert "path"

This reverts commit 8b9395d9cb008e0abbbe199c318b24ff028bd9a6.

* path

* absolute paths

* debug

* :wq
d src/carma-platform
        source /opt/ros/humble/setup.bash
        python -m pytest test --cov=. --cov-report=xml:./coverage.xml --junitxml=./test-results.xml --cov-context=test

* dd debug

* exclude

* works

* typo???

* remove

* debug

* project root

* path

* add to paths

* Another path fix

* unrecognized

* revert to working
e Please enter the commit message for your changes. Lines starting

* clean

* omit some files

* parse_ros2 unit test

* respect the coveragerc

* try agian with ag

* try again

* try agian

* use configuration property

* remove extra files

MishkaMN commented:
@john-chrosniak I added unit tests to the repo. This is so that the scripts are maintained and can be reused. Otherwise, we are creating too many new scripts and duplicating code from previous ROS1 approaches.

MishkaMN committed:
fix sonar smells, reduce cog.complex

MishkaMN commented:
And also had to refactor a bit more to reduce the cognitive complexity raised by the unit test.

MishkaMN committed:
Fix accidental unit test break

MishkaMN committed:
more code smells

john-chrosniak commented on src/carma-platform/test/test_parse_ros2_bags.py:
```diff
@@ -0,0 +1,40 @@
+
+@pytest.fixture
+def mock_mcap_path():
+    return Path("/path/to/mock.mcap")
```
Do we need `mock.mcap` stored in the git repo for these unit tests to run?

MishkaMN replied:
Nope. We don't need it as the `with patch` call just mocks all functions and just returns values expected so we can return any values as if real mcap exists.

john-chrosniak replied:
Got it, thanks for the clarification!
